app/views/shared/views/_prov_dialog.html.haml

```diff
@@ -70,7 +70,7 @@
         :prefix                     => "/miq_request/",
         :keys                       => keys})
   - when :volumes
-    - keys = [:name, :size, :delete_on_terminate]
+    - keys = wf.kind_of?(ManageIQ::Providers::IbmCloud::PowerVirtualServers::CloudManager::ProvisionWorkflow) ? [:name, :size, :diskType, :shareable] : [:name, :size, :delete_on_terminate]
```

Comment:
This is probably a :fire_engine: red flag :fire_engine:, we should avoid conditionals like this :confused:

Reply:
ask, don't tell.  In other words, ask the `wf` object what it supports and render those keys

Reply:
First I was like, what should I ask here :rofl: because I wanted to ask what is the goal with this change :confused: thanks @chessbyte :+1:

Reply:
Hi. Yeah, that's just handling the case of IBM PVS plugin there. In our case we'd have slightly other view on the provisioning form's _Volumes_ tab. In all other cases it should be as it was before. I think there are multiple similar checks in case of other plugins there. E.g. see [here](https://github.com/ManageIQ/manageiq-ui-classic/blob/5c07afc2ef8288fbe3a286055877942eb73aede3/app/views/shared/views/_prov_dialog.html.haml#L33-L35).

Reply:
Well, the example are you pointing at is a bad example and that one should be cleaned up the same way. If we want real pluggable providers, our UI code should not test provider types in conditions at all...

Reply:
Agree that it's not that good to have the types in there, as least due to the fact that UI would break if were to remove our plugin from the Gemfile for some reason. But is there a feasible alternative?

One thing that'd eliminate the type checking would be to define a different "tab" called "new_volumes" that only IBM's YAML would reference but I'm not sure if it'd involve only a new 'when' statement in __prov_dialog.html.haml_. I could check that.

Reply:
You can specify the keys in your provider in a constant/method (just make it declarative-ish) and ask as @chessbyte is suggesting. I am not that much familiar with the area, so I'm not sure what you should name the accessor. As we want to keep compatibility with the rest of the providers without major changes, could define a `keys_for_i_am_unsure_about_the_name_here` method/constant on the root `MiqProvisionWorkflow` class with the original values and you would just override them in your  `ProvisionWorkflow` subclass. Then the line above would be something like:
```haml
- keys = wf.keys_for_i_am_unsure_about_the_name_here
```

Reply:
Not sure about changing such genereic class as "MiqProvisionWorkflow" just for our own provider. At least if the class defined default fields for all existing tabs. Let's ask Adam what he thinks about this change first.

@agrare Do you think it's ok to define a generic method, say 'keys_for_volumes' in MiqProvisionWorkflow?

Reply:
Another less "invasive" way would be perhaps doing something like this in the '_prov_dialog.html.haml'

`- keys = (wf.keys_for_volumes if defined? wf.keys_for_volumes) || default_keys`

This way we could define the fields in our own plugin and not modify the base class.

Reply:
Yeah....`if wf.kind_of?(ManageIQ::Providers::Openstack::CloudManager::ProvisionWorkflow)` this kind of thing is all over this file.

I'd prefer moving the default keys into the base workflow rather than having the defaults here and only overridden in the ibm_cloud repo that way we're consistent with all the providers.

`keys_for_volumes` is a little vague, `keys_for_i_am_unsure_about_the_name_here` works for me...haha

Maybe something that describes what these keys are going to be used for instead of what they are?  `volume_provision_fields` or something like that

Reply:
The first problem of starting a metal band, always hits me, sorry :laughing:

Reply:
:laughing:

Reply:
@chessbyte suggested `volume_dialog_keys` and I can't think of anything better :+1: